lib/synapse/service_watcher/base.rb

```diff
@@ -95,13 +100,44 @@ def validate_discovery_opts
     end
 
     def set_backends(new_backends)
-      if @keep_default_servers
-        @backends = @default_servers + new_backends
+      # Aggregate and deduplicate all potential backend service instances.
+      new_backends = (new_backends + (@keep_default_servers ? @default_servers : [])).uniq {|b|
+        [b['host'], b['port'], b.fetch('name', '')]
+      }
+
+      if new_backends.to_set == @backends.to_set
+        return false
+      end
+
+      if new_backends.empty?
+        if @default_servers.empty?
+          if @use_previous_backends
+            # Discard this update
+            log.warn "synapse: no default servers for service #{@name};" \
+              "  using previous backends: #{@backends.inspect}"
+            return false
+          else
+            log.warn "synapse: no default servers for service #{@name} and" \
+              " 'use_previous_backends' is disabled; dropping all backends"
+            @backends.clear
+          end
+        else
+          log.warn "synapse: no backends for service #{@name};" \
+            " using default servers: #{@default_servers.inspect}"
+          @backends = @default_servers
+        end
       else
+        log.info "synapse: discovered #{new_backends.length} backends for service #{@name}"
```

Comment:
on an unrelated note, one thing i always thought might be necessary eventually is to specify the minimum number of backends to serve traffic.

suppose a service goes down. when functionality is restored, a single backend is naturally going to be the first to come up. that backend might then be pounded with all traffic which is normally meant for many multiple backends, and as a result might go down because of excess traffic. this would then happen to the second backend, etc.. etc.. 

it might be nice here to avoid setting any backends unless the number exceeds some limit which we know can handle the traffic. just a brain-dump, not even necessarily a todo.